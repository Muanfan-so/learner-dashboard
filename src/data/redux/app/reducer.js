import { createSlice } from '@reduxjs/toolkit';

import { StrictDict } from 'utils';

const initialState = {
  pageNumber: 1,
  enrollments: [],
  courseData: {},
  entitlement: [],
  emailConfirmation: {},
  enterpriseDashboard: {},
  platformSettings: {},
  suggestedCourses: [],
  filterState: {},
  selectSessionModal: {},
};

export const cardId = (val) => `card-${val}`;

// eslint-disable-next-line no-unused-vars
const app = createSlice({
  name: 'app',
  initialState,
  reducers: {
    loadCourses: (state, { payload: { courses } }) => ({
      ...state,
      courseData: courses.reduce(
        (obj, curr, index) => ({
          ...obj,
          [cardId(index)]: { ...curr, cardId: cardId(index) },
        }),
        {},
      ),
    }),
    loadGlobalData: (state, { payload }) => ({
      ...state,
      emailConfirmation: payload.emailConfirmation,
      enterpriseDashboard: payload.enterpriseDashboard,
      platformSettings: payload.platformSettings,
      suggestedCourses: payload.suggestedCourses,
    }),
    updateSelectSessionModal: (state, { payload }) => ({
      ...state,
      selectSessionModal: { cardId: payload },
    }),
    setPageNumber: (state, { payload }) => ({ ...state, pageNumber: payload }),
  },
});

const actions = StrictDict(app.actions);

const { reducer } = app;

export {
  actions,
  initialState,
  reducer,
};
