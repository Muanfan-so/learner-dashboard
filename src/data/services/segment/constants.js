import { StrictDict } from 'utils';

export const events = StrictDict({
  courseEnroll: 'courseEnroll',
  entitlementUnenroll: 'entitlementUnenroll',
  sessionChange: 'sessionChange',
  unenrollReason: 'unenrollReason',
  upgradeCourse: 'upgradeCourse',
  searchCourse: 'searchCourse',
});

export const eventNames = StrictDict({
  [events.courseEnroll]: 'edx.bi.user.program-details.enrollment',
  [events.upgradeCourse]: 'learner_home.course_card.upgrade',
  [events.entitlementUnenroll]: 'entitlement_unenrollment_reason.selected',
  [events.sessionChange]: ({ action }) => `course-dashboard.${action}-session`, // 'switch', 'new', 'leave'
  [events.unenrollReason]: 'unenrollment_reason.selected',
  [events.searchCourse]: 'learner_home.widget.search_course',
});

export const trackingCategory = 'learner-home';

export const pageViewEvent = { category: trackingCategory };
