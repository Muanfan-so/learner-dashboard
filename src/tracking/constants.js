import { StrictDict } from 'utils';

export const categories = StrictDict({
  dashboard: 'dashboard',
  upgrade: 'upgrade',
  userEngagement: 'user-engagement',
  searchButton: 'search_button',
});

export const events = StrictDict({
  enterCourseClicked: 'enterCourseClicked',
  courseImageClicked: 'courseImageClicked',
  courseTitleClicked: 'courseTitleClicked',
  courseOptionsDropdownClicked: 'courseOptionsDropdownClicked',
  upgradeButtonClicked: 'upgradeButtonClicked',
  upgradeButtonClickedEnrollment: 'upgradeButtonClickedEnrollment',
  upgradeButtonClickedUpsell: 'upgradeButtonClickedUpsell',
  shareClicked: 'shareClicked',
  userSettingsChanged: 'userSettingsChanged',
  newSession: 'newSession',
  switchSession: 'switchSession',
  leaveSession: 'leaveSession',
  unenrollReason: 'unenrollReason',
  entitlementUnenrollReason: 'entitlementUnenrollReason',
  enterpriseDashboardModalOpened: 'enterpriseDashboardModalOpened',
  enterpriseDashboardModalCTAClicked: 'enterpriseDashboardModalCTAClicked',
  enterpriseDashboardModalClosed: 'enterpriseDashboardModalClosed',
});

const learnerPortal = 'edx.ui.enterprise.lms.dashboard.learner_portal_modal';

export const eventNames = StrictDict({
  enterCourseClicked: 'edx.bi.dashboard.enter_course.clicked',
  courseImageClicked: 'edx.bi.dashboard.course_image.clicked',
  courseTitleClicked: 'edx.bi.dashboard.course_title.clicked',
  courseOptionsDropdownClicked: 'edx.bi.dashboard.course_options_dropdown.clicked',
  upgradeButtonClicked: 'edx.bi.dashboard.upgrade_button.clicked',
  upgradeButtonClickedEnrollment: 'edx.course.enrollment.upgrade.clicked',
  upgradeButtonClickedUpsell: 'edx.bi.ecommerce.upsell_links_clicked',
  shareClicked: 'edx.course.share_clicked',
  userSettingsChanged: 'edx.user.settings.changed',
  newSession: 'course-dashboard.new-session',
  switchSession: 'course-dashboard.switch-session',
  leaveSession: 'course-dashboard.leave-session',
  unenrollReason: 'unenrollment_reason.selected',
  entitlementUnenrollReason: 'entitlement_unenrollment_reason.selected',
  enterpriseDashboardModalOpened: `${learnerPortal}.opened`,
  enterpriseDashboardModalCTAClicked: `${learnerPortal}.dashboard_cta.clicked`,
  enterpriseDashboardModalClosed: `${learnerPortal}.closed`,
  findCoursesClicked: 'edx.bi.dashboard.find_courses_button.clicked',
});

export const appName = 'learner-home';

export default eventNames;
