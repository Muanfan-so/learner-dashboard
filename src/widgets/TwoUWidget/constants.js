/* eslint-disable quotes */
import { StrictDict } from 'utils';

const countryCodeUS = 'US';

const merchandisingKeys = StrictDict({
  execed: 'execed',
  execedDescription: 'execedDescription',
  masters: 'masters',
  mastersDescription: 'mastersDescription',
  bachelors: 'bachelors',
  bachelorsDescription: 'bachelorsDescription',
  bootcamps: 'bootcamps',
  bootcampsDescription: 'bootcampsDescription',
});

const messages = StrictDict({
  [merchandisingKeys.execed]: {
    id: 'TwoUWidget.execed',
    defaultMessage: 'Executive Education',
    description: 'The title for the link to the executive education page.',
  },
  [merchandisingKeys.execedDescription]: {
    id: 'TwoUWidget.execedDescription',
    defaultMessage: 'Short courses to develop leadership skills',
    description: 'The description for the link to the executive education page.',
  },
  [merchandisingKeys.masters]: {
    id: 'TwoUWidget.masters',
    defaultMessage: `Master's Degrees`,
    description: 'The title for the link to the masters page.',
  },
  [merchandisingKeys.mastersDescription]: {
    id: 'TwoUWidget.mastersDescription',
    defaultMessage: 'Online degree programs from top universities',
    description: 'The description for the link to the masters page.',
  },
  [merchandisingKeys.bachelors]: {
    id: 'TwoUWidget.bachelors',
    defaultMessage: `Bachelor's Degrees`,
    description: 'The title for the link to the bachelors page.',
  },
  [merchandisingKeys.bachelorsDescription]: {
    id: 'TwoUWidget.bachelorsDescription',
    defaultMessage: 'Begin or complete a degree; fully online',
    description: 'The description for the link to the bachelors page.',
  },
  [merchandisingKeys.bootcamps]: {
    id: 'TwoUWidget.bootcamps',
    defaultMessage: 'Boot Camps',
    description: 'The title for the link to the boot camps page.',
  },
  [merchandisingKeys.bootcampsDescription]: {
    id: 'TwoUWidget.bootcampsDescription',
    defaultMessage: 'Intensive, hands-on, project-based training',
    description: 'The description for the link to the boot camps page.',
  },
});

const merchandisingItems = [
  {
    key: merchandisingKeys.execed,
    title: messages[merchandisingKeys.execed],
    description: messages[merchandisingKeys.execedDescription],
    itemURL: 'https://www.edx.org/executive-education?vanguards_click=execed',
  },
  {
    key: merchandisingKeys.masters,
    title: messages[merchandisingKeys.masters],
    description: messages[merchandisingKeys.mastersDescription],
    itemURL: 'https://www.edx.org/masters?vanguards_click=masters',
  },
  {
    key: merchandisingKeys.bachelors,
    title: messages[merchandisingKeys.bachelors],
    description: messages[merchandisingKeys.bachelorsDescription],
    itemURL: 'https://www.edx.org/bachelors?vanguards_click=bachelors',
  },
  {
    key: merchandisingKeys.bootcamps,
    title: messages[merchandisingKeys.bootcamps],
    description: messages[merchandisingKeys.bootcampsDescription],
    itemURL: 'https://www.edx.org/boot-camps?vanguards_click=bootcamps',
  },
];

const getMerchandisingItems = (countryCode) => (
  countryCode === countryCodeUS ? merchandisingItems : merchandisingItems.slice(0, 3)
);

export { getMerchandisingItems, countryCodeUS };
