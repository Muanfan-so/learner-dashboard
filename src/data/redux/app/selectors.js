import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { FilterKeys } from 'data/constants/app';
import urls from 'data/services/lms/urls';

import * as module from './selectors';

const { baseAppUrl, learningMfeUrl } = urls;

export const appSelector = (state) => state.app;

const mkSimpleSelector = (cb) => createSelector([module.appSelector], cb);

// top-level app data selectors
export const simpleSelectors = {
  courseData: mkSimpleSelector(app => app.courseData),
  platformSettings: mkSimpleSelector(app => app.platformSettings),
  suggestedCourses: mkSimpleSelector(app => app.suggestedCourses),
  emailConfirmation: mkSimpleSelector(app => app.emailConfirmation),
  enterpriseDashboard: mkSimpleSelector(app => app.enterpriseDashboard),
  selectSessionModal: mkSimpleSelector(app => app.selectSessionModal),
  pageNumber: mkSimpleSelector(app => app.pageNumber),
};

export const numCourses = createSelector(
  [module.simpleSelectors.courseData],
  (courseData) => Object.keys(courseData).length,
);
export const hasCourses = createSelector([module.numCourses], (num) => num > 0);
export const hasAvailableDashboards = createSelector(
  [module.simpleSelectors.enterpriseDashboard],
  (data) => data !== null,
);
export const showSelectSessionModal = createSelector(
  [module.simpleSelectors.selectSessionModal],
  (data) => data.cardId != null,
);

export const courseCardData = (state, cardId) => (
  module.simpleSelectors.courseData(state)[cardId]
);

const mkCardSelector = (sel) => (state, cardId) => {
  const cardData = module.courseCardData(state, cardId);
  return sel(cardData);
};

const dateSixMonthsFromNow = new Date();
dateSixMonthsFromNow.setDate(dateSixMonthsFromNow.getDate() + 180);
const today = new Date();

export const courseCard = StrictDict({
  certificate: mkCardSelector(({ certificate }) => {
    const availableDate = certificate.availableDate ? new Date(certificate.availableDate) : null;

    return {
      availableDate,
      certPreviewUrl: certificate.certPreviewUrl,
      isDownloadable: certificate.isDownloadable,
      isEarnedButUnavailable: certificate.isEarned && availableDate > today,
      isRestricted: certificate.isRestricted,
    };
  }),
  course: mkCardSelector(({ course }) => ({
    bannerImgSrc: baseAppUrl(course.bannerImgSrc),
    courseNumber: course.courseNumber,
    courseName: course.courseName,
    website: course.website,
  })),
  courseRun: mkCardSelector(({ courseRun }) => (courseRun === null ? {} : {
    endDate: courseRun.endDate ? new Date(courseRun?.endDate) : null,
    courseId: courseRun.courseId,
    isArchived: courseRun.isArchived,
    isStarted: courseRun.isStarted,
    isFinished: courseRun.isFinished,
    minPassingGrade: Math.floor(courseRun.minPassingGrade * 100),
    startDate: courseRun.startDate ? new Date(courseRun.startDate) : null,
    homeUrl: courseRun.homeUrl,
    marketingUrl: courseRun.marketingUrl,
    progressUrl: learningMfeUrl(courseRun.progressUrl),
    unenrollUrl: learningMfeUrl(courseRun.unenrollUrl),
    upgradeUrl: courseRun.upgradeUrl,
    resumeUrl: learningMfeUrl(courseRun.resumeUrl),
  })),
  enrollment: mkCardSelector(({ enrollment }) => {
    if (enrollment == null) {
      return {
        isEnrolled: false,
      };
    }
    const { isStaff, hasUnmetPrereqs, isTooEarly } = enrollment.coursewareAccess;
    return {
      accessExpirationDate: enrollment.accessExpirationDate ? new Date(enrollment.accessExpirationDate) : null,
      canUpgrade: enrollment.canUpgrade,
      hasStarted: enrollment.hasStarted,
      coursewareAccess: enrollment.coursewareAccess,
      hasAccess: isStaff || !(hasUnmetPrereqs || isTooEarly),
      hasFinished: enrollment.hasFinished,
      isAudit: enrollment.isAudit,
      isAuditAccessExpired: enrollment.isAuditAccessExpired,
      isEmailEnabled: enrollment.isEmailEnabled,
      isVerified: enrollment.isVerified,
      lastEnrolled: enrollment.lastEnrollment,
      isEnrolled: enrollment.isEnrolled,
    };
  }),
  entitlement: mkCardSelector(({ entitlement }) => {
    if (!entitlement || Object.keys(entitlement).length === 0) {
      return { isEntitlement: false };
    }
    const deadline = new Date(entitlement.changeDeadline);
    const deadlinePassed = deadline < new Date();
    const showExpirationWarning = !deadlinePassed && deadline <= dateSixMonthsFromNow;
    return {
      canChange: !deadlinePassed,
      entitlementSessions: entitlement.availableSessions,
      isEntitlement: true,
      isExpired: entitlement.isExpired,
      isFulfilled: entitlement.isFulfilled,
      hasSessions: entitlement.availableSessions?.length > 0,
      changeDeadline: deadline,
      uuid: entitlement.uuid,
      showExpirationWarning,
    };
  }),
  gradeData: mkCardSelector(({ gradeData }) => ({ isPassing: gradeData.isPassing })),
  courseProvider: mkCardSelector(({ courseProvider }) => ({ name: courseProvider?.name })),
  relatedPrograms: mkCardSelector(({ programs: { relatedPrograms } }) => ({
    list: relatedPrograms.map(program => ({
      bannerImgSrc: program.bannerImgSrc,
      logoImgSrc: program.logoImgSrc,
      numberOfCourses: program.numberOfCourses,
      programType: program.programType,
      programUrl: program.programUrl,
      provider: program.provider,
      title: program.title,
    })),
    length: relatedPrograms.length,
  })),
});

export const currentList = (state, {
  sortBy,
  isAscending,
  filters,
  pageSize,
}) => {
  const pageNumber = module.simpleSelectors.pageNumber(state);
  let list = Object.values(module.simpleSelectors.courseData(state));
  const hasFilter = filters.reduce((obj, filter) => ({ ...obj, [filter]: true }), {});
  if (filters.length) {
    list = list.filter(course => {
      if (hasFilter[FilterKeys.notEnrolled]) {
        if (course.enrollment.isEnrolled) {
          return false;
        }
      }
      if (hasFilter[FilterKeys.done]) {
        if (!course.enrollment.hasFinished) {
          return false;
        }
      }
      if (hasFilter[FilterKeys.upgraded]) {
        if (!course.enrollment.isVerified) {
          return false;
        }
      }
      if (hasFilter[FilterKeys.inProgress]) {
        if (!course.enrollment.hasStarted) {
          return false;
        }
      }
      if (hasFilter[FilterKeys.notStarted]) {
        if (course.enrollment.hasStarted) {
          return false;
        }
      }
      return true;
    });
  }
  if (sortBy === 'enrolled') {
    list = list.sort((a, b) => {
      const dateA = new Date(a.enrollment.lastEnrolled);
      const dateB = new Date(b.enrollment.lastEnrolled);
      if (dateA < dateB) { return isAscending ? -1 : 1; }
      if (dateA > dateB) { return isAscending ? 1 : 1; }
      return 0;
    });
  } else {
    list = list.sort((a, b) => {
      const nameA = a.course.courseName.toLowerCase();
      const nameB = b.course.courseName.toLowerCase();
      if (nameA < nameB) { return isAscending ? -1 : 1; }
      if (nameA > nameB) { return isAscending ? 1 : 1; }
      return 0;
    });
  }
  return {
    visible: list.slice((pageNumber - 1) * pageSize, pageNumber * pageSize),
    numPages: Math.ceil(list.length / pageSize),
  };
};

export default StrictDict({
  ...simpleSelectors,
  courseCard,
  currentList,
  hasCourses,
  hasAvailableDashboards,
  showSelectSessionModal,
});
