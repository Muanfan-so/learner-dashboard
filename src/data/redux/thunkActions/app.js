import { StrictDict } from 'utils';
import { handleEvent } from 'data/services/segment/utils';
import { eventNames } from 'data/services/segment/constants';
import { actions, selectors } from 'data/redux';
import { post } from 'data/services/lms/utils';

import requests from './requests';

// import { locationId } from 'data/constants/app';

// import { } from './requests';
import * as module from './app';

export const loadData = ({ courses, ...globalData }) => dispatch => {
  dispatch(actions.app.setPageNumber(1));
  dispatch(actions.app.loadGlobalData(globalData));
  dispatch(actions.app.loadCourses({ courses }));
};

/**
 * initialize the app, loading ora and course metadata from the api, and loading the initial
 * submission list data.
 */
export const initialize = () => (dispatch) => (
  dispatch(requests.initializeList({
    onSuccess: (response) => dispatch(module.loadData(response)),
  }))
);

export const refreshList = () => (dispatch) => (
  dispatch(requests.initializeList({
    onSuccess: (response) => dispatch(module.loadData(response)),
  }))
);

// TODO: connect hook to actual api later
export const sendConfirmEmail = () => (dispatch, getState) => post(
  selectors.app.emailConfirmation(getState()).sendEmailUrl,
);

export const newEntitlementEnrollment = (cardId, selection) => (dispatch, getState) => {
  const { uuid } = selectors.app.courseCard.entitlement(getState(), cardId);
  handleEvent(eventNames.sessionChange({ action: 'new' }), {
    fromCourseRun: null,
    toCourseRun: selection,
  });
  return dispatch(requests.newEntitlementEnrollment({ uuid, courseId: selection }));
};

export const switchEntitlementEnrollment = (cardId, selection) => (dispatch, getState) => {
  const { courseId } = selectors.app.courseCard.courseRun(getState(), cardId);
  const { uuid } = selectors.app.courseCard.entitlement(getState(), cardId);
  handleEvent(eventNames.sessionChange({ action: 'switch' }), {
    fromCourseRun: courseId,
    toCourseRun: selection,
  });
  return dispatch(requests.switchEntitlementEnrollment({ uuid, courseId: selection }));
};

export const leaveEntitlementSession = (cardId) => (dispatch, getState) => {
  const { courseId } = selectors.app.courseCard.courseRun(getState(), cardId);
  const { uuid } = selectors.app.courseCard.entitlement(getState(), cardId);
  handleEvent(eventNames.entitlementUnenroll({ action: 'leave' }), {
    fromCourseRun: courseId,
    toCourseRun: null,
  });
  return dispatch(requests.leaveEntitlementSession({ uuid }));
};

export const unenrollFromCourse = (courseId, reason) => (dispatch) => {
  handleEvent(eventNames.unenrollReason, {
    category: 'user-engagement',
    displayName: 'v1',
    label: reason,
    course_id: courseId,
  });
  dispatch(requests.unenrollFromCourse({
    courseId,
    onSuccess: () => dispatch(module.refreshList()),
  }));
};

export const masqueradeAs = (user) => (dispatch) => (
  dispatch(requests.masqueradeAs({
    user,
    onSuccess: (({ courses }) => {
      dispatch(actions.app.loadCourses({ courses }));
    }),
  }))
);

export const clearMasquerade = () => (dispatch) => {
  dispatch(requests.clearMasquerade());
  dispatch(module.refreshList());
};

export default StrictDict({
  loadData,
  initialize,
  refreshList,
  sendConfirmEmail,
  newEntitlementEnrollment,
  switchEntitlementEnrollment,
  leaveEntitlementSession,
  unenrollFromCourse,
  masqueradeAs,
  clearMasquerade,
});
