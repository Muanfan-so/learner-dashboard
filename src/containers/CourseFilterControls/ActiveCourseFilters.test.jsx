import { shallow } from 'enzyme';

import ActiveCourseFilters from './ActiveCourseFilters';

import { FilterKeys } from 'data/constants/app';

describe('ActiveCourseFilters', () => {
  const props = {
    filters: Object.values(FilterKeys),
    setFilters: {
      remove: jest.fn().mockName('setFilters.remove'),
      clear: jest.fn().mockName('setFilters.clear'),
    },
    handleRemoveFilter: jest.fn().mockName('handleRemoveFilter'),
  };
  describe('snapshot', () => {
    it('renders', () => {
      const wrapper = shallow(<ActiveCourseFilters {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});