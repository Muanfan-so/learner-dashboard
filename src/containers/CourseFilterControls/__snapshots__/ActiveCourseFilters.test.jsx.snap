// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveCourseFilters snapshot renders 1`] = `
<div
  id="course-list-active-filters"
>
  <Chip
    key="inProgress"
    variant="primary"
  >
    In-Progress
  </Chip>
  <Chip
    key="notStarted"
    variant="primary"
  >
    Not Started
  </Chip>
  <Chip
    key="done"
    variant="primary"
  >
    Done
  </Chip>
  <Chip
    key="notEnrolled"
    variant="primary"
  >
    Not Enrolled
  </Chip>
  <Chip
    key="upgraded"
    variant="primary"
  >
    Upgraded
  </Chip>
  <Button
    onClick={[MockFunction setFilters.clear]}
    variant="link"
  >
    Clear all
  </Button>
</div>
`;
